### PR TITLE
ENT-3430: Add a script to run cf-serverd tests with various versions

### PR DESCRIPTION
**Pull request by Lex-2008**
And add this as a new job to Travis.

These versions are tested so far:
* currently built server with 3.10.2 agent
* currently built server with 3.7.6 agent
* currently built agent with 3.10.2 server
* currently built agent with 3.7.6 server

### CODE REVIEW

File: tests/acceptance/serverd-multi-versions.sh [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+# Ensure reverse hostname resolution is correct and 127.0.0.1 is always 'localhost'.
+# There's no nice shell command to test it but this one:
+# python -c 'import socket;print socket.gethostbyaddr("127.0.0.1")'
+sudo sed -i -e '1s/^/127.0.0.1 localhost localhost.localdomian\n/' /etc/hosts
+
+# downloads $3 (if provided), saves it as $2 and extracts it to $1
+# if $2 not provided - installs currently built version
+function install(){
+    test -z "$1" && return 1
+    rm -rf $PWD/$1
```
Comment (jimis) [inline tests/acceptance/serverd-multi-versions.sh]: Why do you prepend `$PWD` to the arguments? Isn't `rm -rf "$1"` working?

EDIT: I'm a bit confused by the $PWDs all over. Why?

Comment (jimis) [reply tests/acceptance/serverd-multi-versions.sh]: Also use double quotes around the expansions with `rm -rf` for safety.

File: tests/acceptance/serverd-multi-versions.sh [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		mv work_here $logfile.workdir
+		mv /tmp/TESTDIR.cfengine $logfile.testdir
+	fi
+    done
+}
+
+install hub
+install agent 3.10.deb https://cfengine-package-repos.s3.amazonaws.com/community_binaries/Community-3.10.2/agent_debian7_x86_64/cfengine-community_3.10.2-1_amd64-debian7.deb
+testit master-hub-with-3.10.2-agent
+install agent 3.7.deb https://cfengine-package-repos.s3.amazonaws.com/community_binaries/Community-3.7.6/agent_deb_x86_64/cfengine-community_3.7.6-1_amd64-debian4.deb
```
Comment (jimis) [inline tests/acceptance/serverd-multi-versions.sh]: Leave a blank line after each `testit` line for readability.

File: tests/acceptance/serverd-multi-versions.sh [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+function testit(){
+    echo "===== $1"
+    $PWD/server.sh -V
+    $PWD/agent.sh -V
+    ls tests/acceptance/16_cf-serverd/serial/*.cf | grep -v -f tests/acceptance/serverd-multi-versions-blacklist.txt | sed 's%^tests/%%' | while read file
+    do
+        logfilename="$1-$(basename $file)"
+        logfile="$logdir/$logfilename"
```
Comment (jimis) [inline tests/acceptance/serverd-multi-versions.sh]: I **love** that the logs are so easily accessible! Looking forward to seeing this in the classic acceptance tests.

Comment (jimis) [reply tests/acceptance/serverd-multi-versions.sh]: On another note, the logs include only the client, and include messages like the following:

> R: /home/travis/build/cfengine/core/server.sh was run in debug mode, the logs will be in /home/travis/build/cfengine/core/work_here/server-debug.localhost_open.srv.log

Can we include the serverd logs in the archive?

Comment (jimis) [reply tests/acceptance/serverd-multi-versions.sh]: And on another note, the test passes, but I see complaints of "abnormal termination":

> Q: "...e/agent.sh -KIf": R: /home/travis/build/cfengine/core/work_here/inputs/acceptance/16_cf-serverd/serial/001.cf Pass
>    error: Finished script '/home/travis/build/cfengine/core/server.sh -Kdf /home/travis/build/cfengine/core/work_here/inputs/acceptance/16_cf-serverd/serial/localhost_open.srv > /home/travis/build/cfengine/core/work_here/server-debug.localhost_open.srv.log 2>&1' - failed (abnormal termination)

Any ideas why?

File: tests/acceptance/serverd-multi-versions.sh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        $PWD/server.sh -V >> $logfile.log
+        $PWD/agent.sh -V >> $logfile.log
+        until rm -rf /tmp/TESTDIR.cfengine
```
Comment (jimis) [inline tests/acceptance/serverd-multi-versions.sh]: why do you loop? one line comment would be nice

File: tests/acceptance/serverd-multi-versions-blacklist.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Note: this file must NOT contain empty lines, or all tests will be blacklisted!
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions-blacklist.txt]: but they're testing _something_, don't they?

File: tests/acceptance/serverd-multi-versions.sh [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+rm -rf work_stub/inputs/acceptance/0*
+mkdir -p work_stub/bin/
+touch work_stub/bin/cf-promises
+chmod a+x work_stub/bin/cf-promises
+
+sudo hostname localhost
+# Ensure reverse hostname resolution is correct and 127.0.0.1 is always 'localhost'.
+# There's no nice shell command to test it but this one:
+# python -c 'import socket;print socket.gethostbyaddr("127.0.0.1")'
+sudo sed -i -e '1s/^/127.0.0.1 localhost localhost.localdomian\n/' /etc/hosts
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: is there something wrong with it?

File: tests/acceptance/serverd-multi-versions.sh [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+function install(){
+    test -z "$1" && return 1
+    rm -rf $PWD/$1
+    if test -z "$2"
+    then
+        make --silent DESTDIR=$PWD/$1 install
+        mkdir -p $PWD/$1/var
+        mv $PWD/$1/$HOME/cf_install $PWD/$1/var/cfengine
+    else
+        test -z "$3" -o -f $2 || wget $3 -O $2
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: hm, you're right, maybe two negatives are more readable than one...
```
test ! -z "$3" -a ! -f "$2" && wget "$3" -O "$2"
```

File: tests/acceptance/serverd-multi-versions.sh (hunk 3 above)
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: serverd logs are included in *.workdir dir in archive for failed tests. I thought that it's a waste to include them for passed tests.

Regarding `error: Finished script '...' - failed (abnormal termination)` - it looks like expected CFEngine output. When we terminate cf-serverd, we send two signals to same processes: SIGTERM and SIGKILL [(run_with_server.cf.sub#L106)](https://github.com/cfengine/core/blob/master/tests/acceptance/run_with_server.cf.sub#L106). If server shuts down cleanly after first signal - the second doesn't reach its target and prints error message like "process PID not found"; if server does not shut down nicely after first signal - it gets killed and above error message gets printed. That's my understanding, at least.

File: tests/acceptance/serverd-multi-versions.sh [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    $PWD/server.sh -V
+    $PWD/agent.sh -V
+    ls tests/acceptance/16_cf-serverd/serial/*.cf | grep -v -f tests/acceptance/serverd-multi-versions-blacklist.txt | sed 's%^tests/%%' | while read file
+    do
+        logfilename="$1-$(basename $file)"
+        logfile="$logdir/$logfilename"
+        $PWD/server.sh -V >> $logfile.log
+        $PWD/agent.sh -V >> $logfile.log
+        until rm -rf /tmp/TESTDIR.cfengine
+	do
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: dunno how it slipped in

File: tests/acceptance/serverd-multi-versions.sh (hunk 1 above)
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: right.

**Comment by Lex-2008** [comment]
@jimis : PR updated with requested changes, please re-review!

File: tests/acceptance/serverd-multi-versions-blacklist.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (jimis) [reply tests/acceptance/serverd-multi-versions-blacklist.txt]: They are not testing anything related to cross-version compatibility. They test the syntax-checker, which is checked in the normal acceptance tests run.

File: tests/acceptance/serverd-multi-versions.sh (hunk 3 above)
Comment (jimis) [reply tests/acceptance/serverd-multi-versions.sh]: > Regarding error: Finished script '...' - failed (abnormal termination) - it looks like expected CFEngine output. When we terminate cf-serverd, we send two signals to same processes: SIGTERM and SIGKILL (run_with_server.cf.sub#L106). If server shuts down cleanly after first signal - the second doesn't reach its target and prints error message like "process PID not found"; if server does not shut down nicely after first signal - it gets killed and above error message gets printed. That's my understanding, at least.

Ugly, but you are right. I have always wanted to insert a delay between the 2 signals: https://tracker.mender.io/browse/CFE-2207

File: tests/acceptance/serverd-multi-versions.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (jimis) [reply tests/acceptance/serverd-multi-versions.sh]: You hidden tab supporter. THOU SHALL NOT PASS!

File: tests/acceptance/serverd-multi-versions.sh [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+sudo sed -i -e '1s/^/127.0.0.1 localhost localhost.localdomain\n/' /etc/hosts
+
+# downloads $3 (if provided), saves it as $2 and extracts it to $1
+# if $2 not provided - installs currently built version
+function install(){
+    test -z "$1" && return 1
+    rm -rf "$1"
+    if test -z "$2"
+    then
+        make install prefix=$PWD/$1/var/cfengine 2>&1 >/dev/null
```
Comment (jimis) [inline tests/acceptance/serverd-multi-versions.sh]: `make install` fails in the logs, it looks like a bug in our makefiles, they try to install to system path regardless of prefix!

File: tests/acceptance/serverd-multi-versions.sh [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    do
+        logfilename="$1-$(basename $file)"
+        logfile="$logdir/$logfilename"
+        ./server.sh -V >> $logfile.log
+        ./agent.sh -V >> $logfile.log
+        # delete dirs left here from previous testing round, if any.
+        # We keep deleting them until they disappear - they can be kept back by
+        # still-running process from previous testing round.
+        while test -d /tmp/TESTDIR.cfengine
+	do
```
Comment (jimis) [inline tests/acceptance/serverd-multi-versions.sh]: tab

File: tests/acceptance/serverd-multi-versions.sh (hunk 9 above)
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: Right. I've created https://tracker.mender.io/browse/CFE-2738 but I'm not sure what's the correct location would be in this case

File: tests/acceptance/serverd-multi-versions.sh (hunk 10 above)
Comment (Lex-2008) [reply tests/acceptance/serverd-multi-versions.sh]: damnit!

**Pull request merged**